Utils.ps1

```diff
@@ -104,6 +140,281 @@ function Set-LabAccessControl {
+          $allocatedSubnetRanges[$subnetAddressRangeStart] = $nextsubnetAddressRangeStart
+        }
+        else {
+          $allocatedSubnetRanges.Add($subnetAddressRangeStart, $nextsubnetAddressRangeStart)
+        }
+      }
+    }
+
+    # Ordered Array of unallocated ranges (CIDR notation) calculated as a complement to the above table
+    $unallocatedSubnetRanges = [System.Collections.ArrayList]@()
+
+    $allocatedSubnetStarts = [array]$allocatedSubnetRanges.Keys
+    $allocatedSubnetStarts | ForEach-Object {
+
+      $index = $allocatedSubnetStarts.IndexOf($_)
+      $nextsubnetAddressRangeStart = $allocatedSubnetRanges[$_]
+      if ($index+1 -lt $allocatedSubnetRanges.Count) {
+
+        $nextNextsubnetAddressRangeStart = $allocatedSubnetStarts[$index+1]
+        if ($nextsubnetAddressRangeStart -ne $nextNextsubnetAddressRangeStart) {
+          
+          # Calculate the end IP of the allocated range
+          $nextsubnetAddressRangeEnd = (($nextNextsubnetAddressRangeStart | Convert-IPToDecimal) - 1) | Convert-DecimalToIp 
+          
+          $unallocatedAddressRange = Convert-IPRangeToCIDRNotation -Start $nextsubnetAddressRangeStart -End $nextsubnetAddressRangeEnd 
+          $unallocatedSubnetRanges.Add($unallocatedAddressRange) | Out-Null
+
+        } # else contiguous subnet
+      }
+    }
+    
+    # TODO deciding on the smallest (bottom) or shortest subnet
+    $unallocatedSubnetRanges | Select-Object -First 1
+  }
+  catch {
+    Write-Error -ErrorRecord $_
+  }
+  finally {
+    Remove-PsipcalcScript
+  }
+}
+
+# Get the first available subnet with a unassigned address space large enough to host a new subnet of at least /$Length
+function Get-VirtualNetworkUnassignedSpace {
+  [CmdletBinding()]
+  param(    
+    [parameter(Mandatory=$true,HelpMessage="Virtual Network to get the unassigned space from")]
+    [ValidateNotNullOrEmpty()]
+    $VirtualNetwork,
+
+    [parameter(Mandatory=$true,HelpMessage="Length of the required address space")]
+    [ValidateNotNullOrEmpty()]
+    [int] $Length
+  )
+
+  try {
+
+    # Using Psipcal
+    Import-PsipcalcScript
+
+    # Get Highest address prefix length
+    $networkLengths = $VirtualNetwork.AddressSpace.AddressPrefixes | Select-Object -Property @{Name = 'NetworkLength'; Expression = {$_.Split("/")[1]}} | Select-Object -ExpandProperty NetworkLength
+    $highestNetworkLength = [Linq.Enumerable]::Min([int[]] $networkLengths)
+    if ($highestNetworkLength -ge $Length) {
+      throw "You must use a Vnet of at least /$($Length+1) or larger"
+    }
+
+    # Make sure to get the 'subnets/ipConfigurations' of the underlying VNet
+    $VirtualNetwork = Get-AzVirtualNetwork -ResourceGroupName $VirtualNetwork.ResourceGroupName -Name $VirtualNetwork.Name -ExpandResource 'subnets/ipConfigurations'
+
+    # Using foreach instead of Foreach-Object to break execution of the loop without exiting the function
+    foreach ($_ in $VirtualNetwork.Subnets  | `
```

Comment:
For each subnet:

1. We get the min and max assigned IP
2. We check whether these sit in the lower or upper half of the subnet space
3. If yes, we break from the statement